ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs

```diff
@@ -764,6 +766,69 @@ runWithTempRegistry m = withRegistry $ \rr -> do
         }
     return a
 
+-- | Embed a self-contained 'WithTempRegistry' computation into a larger one.
+--
+-- The internal 'WithTempRegistry' is effectively passed to
+-- 'runWithTempRegistry'. It therefore must have no dangling resources, for
+-- example. This is the meaning of /self-contained/ above.
+--
+-- The key difference beyond 'runWithTempRegistry' is that the resulting
+-- composite resource is also guaranteed to be registered in the outer
+-- 'WithTempRegistry' computation's registry once the inner registry is closed.
+-- Combined with the following assumption, this establishes the invariant that
+-- all resources are (transitively) in a temporary registry.
+--
+-- As the resource might require some implementation details to be closed, the
+-- function to close it will also be provided by the inner computation.
+--
+-- ASSUMPTION: closing @res@ closes every resource contained in @inner_st@
+--
+-- NOTE: In the current implementation, there will be a brief moment where the
+-- inner registry still contains the inner computation's resources and also the
+-- outer registry simultaneously contains the new composite resource. If an async
+-- is received at that time, then the inner resources will be closed and then
+-- the composite resource will be later closed. This means there's a risk of
+-- /double freeing/, which can be harmless if anticipated.
+runInnerWithTempRegistry
+  :: forall inner_st st m res a. IOLike m
+  => WithTempRegistry inner_st m (a, inner_st, res, res -> m Bool)
+     -- ^ The embedded computation; see ASSUMPTION above
+  -> (st -> res -> Bool)
+     -- ^ How to check; same as for 'allocateTemp'
+  -> WithTempRegistry st m a
+runInnerWithTempRegistry inner isTransferred = do
+    outerTR <- WithTempRegistry ask
+
+    lift $ runWithTempRegistry $ do
+      (a, inner_st, res, free) <- inner
+
```

Comment:
What happens if an async exception is thrown at this point? Aren't we leaking `res` since it won't be transferred yet to the outer registry?

Reply:
Hmm. That's a good question.

The ASSUMPTION we came with up so far is that "closing `res` must close `inner_st`". But your observation here seems dual to that assumption: what about the possibility of stuff in `res` that is not also in `inner_st`?

For our current use cases, if I understand correctly, any such "stuff" (eg threads?) is being allocated in the top-level registry. So no resource is _only_ tracked in `res` (and not in `inner_st`). Is that right, @Jasagredo?

Reply:
Supposedly, the inner registry is not yet vanished and in presence of an exception it should be able to close any open resources on the values it returned. And the assumption is that keeping track or `res` we can close it from the outer registry and this should be good enough. As you said, no resource is `only` tracked in `res` and not in `inner_st`, but even if it was, the inner temp registry should take care of it.

Reply:
[This is a "review comment", but it's in a thread in `ResourceRegistry.hs`.]

> As you said, no resource is _only_ tracked in `res` and not in `inner_st`, but even if it was, the inner temp registry should take care of it.

Yeah, if the `inner` computation allocates all of its resources with `allocateTemp`, then anything not in `innerSt` will be reaped when the inner registry is called (and will presumably cause an exception since it isn't in `innerSt`; we are calling `runWithTempRegistry` here for the inner computation!).

Javier: would you add a comment just above the `pure (a, innerSt)` line explaining that? Something like:

> Note that everything `inner` allocated via `allocateTemp` must either be closed or else present in `innerSt` by this point -- `runWithTempRegistry` would have thrown if not.

Reply:
Yes, I just added one :ok_hand: